packages/devtools_app/lib/src/app.dart

```diff
@@ -315,17 +316,35 @@ class DevToolsScreen<C> {
   /// widgets depending on this controller can access it by calling
   /// `Provider<C>.of(context)`.
   ///
-  /// If null, [screen] will be responsible for creating and maintaining its own
-  /// controller.
+  /// If [createController] and [controller] are both null, [screen] will be
+  /// responsible for creating and maintaining its own controller.
   final C Function() createController;
 
+  /// A provided controller for this screen, if non-null.
+  ///
+  /// The controller will then be provided via [controllerProvider], and
+  /// widgets depending on this controller can access it by calling
+  /// `Provider<C>.of(context)`.
+  ///
+  /// If [createController] and [controller] are both null, [screen] will be
+  /// responsible for creating and maintaining its own controller.
+  final C controller;
+
+  /// Returns true if a controller was provided for [screen]. If false,
+  /// [screen] is responsible for creating and maintaining its own controller.
+  bool get providesController => createController != null || controller != null;
+
   /// Whether this screen has implemented offline support.
   ///
   /// Defaults to false.
   final bool supportsOffline;
 
   Provider<C> get controllerProvider {
-    assert(createController != null);
+    assert((createController != null && controller == null) ||
+        (createController == null && controller != null));
```

Comment:
Actually, I think we should allow both to be null based on the comment above `providesController`

Reply:
so this can just be `assert(createController == null || controller == null)` and instead of doing this here, do this at the constructor.

Reply:
Should we be calling `controllerProvider` when there's no controller provided? In that case there's nothing to create a `Provider` for.

Reply:
good point. We can leave this as is then. I still think adding `assert(createController == null || controller == null)` at the constructor level would be a good idea so it is clear from the start that you should not be providing both a `controller` and `createController` param.

Reply:
Done.